### DISCUSSION TITLE
Change stdout and stderr log access modes

### DISCUSSION THREAD

**Discussion by zo-zoyo** [apache/airflow#25182] [General]
Starting Airflow components using the `--stdout` and `--stderr` arguments redirects the output streams to a log file. Currently the access type for these log files is `w+`, as you can see below (might not include all affected components).

Scheduler:
https://github.com/apache/airflow/blob/de6938e173773d88bd741e43c7b0aa16d8a1a167/airflow/cli/commands/scheduler_command.py#L68

Flower:
https://github.com/apache/airflow/blob/de6938e173773d88bd741e43c7b0aa16d8a1a167/airflow/cli/commands/celery_command.py#L70

Celery:
https://github.com/apache/airflow/blob/de6938e173773d88bd741e43c7b0aa16d8a1a167/airflow/cli/commands/celery_command.py#L180

Webserver:
https://github.com/apache/airflow/blob/de6938e173773d88bd741e43c7b0aa16d8a1a167/airflow/cli/commands/webserver_command.py#L452

Triggerer:
https://github.com/apache/airflow/blob/de6938e173773d88bd741e43c7b0aa16d8a1a167/airflow/cli/commands/triggerer_command.py#L42

My problem with this access mode is that it can create sparse files. If the log file is moved or deleted, a new one will be created and the stdout/stderr will be written into it, but not from the start of the file. Instead if the original log file was `x` characters long, the new log file will contain `x` null bytes, and the new lines will be appended after that.

This behaviour prevents us from rotating the log files properly. The only solution for it currently without changing the code is restarting the components after each log rotation, which is not really feasible.

I propose changing the access mode for these output files from `w+` to `a`. As I've seen it from my tests it still creates the log file if it doesn't exist, but it does not print null bytes at the beginning of new files if the old one is deleted.

The problem with this solution is that old output files are not automatically deleted (or, to put it better, overridden) at component start, which might result in very big output files in the long run. To preserve this behaviour, we could do something like this after opening the file:

```python
with open(stdout, "a") as stdout, open(stderr, "a") as stderr:
    stdout.seek(0)
    stdout.truncate()
    stderr.seek(0)
    stderr.truncate()
```

**Comment by potiuk** [comment]
Why don't you provide a PR fixing it ? I think the solution with "append" mode and truncation sounds lke a good idea.

**Reply by enoperm** [reply]
Are you sure about the truncation part? This issue arises on rotation, which I believe is in itself a common solution to log files getting big in the long run, and unlike truncating on open, it shouldn't lose any already written log records. I think simply switching to append mode in itself would be a nicer solution.

**Reply by enoperm** [reply]
That said, I have played around with a simplified version of the code, and indeed, the current version truncates, changing it may cause unforeseen problems in existing deployments without log rotation.
```python
#!/usr/bin/env python3

import daemon
import time

with open('stdout.log', 'w+') as stdout_handle:
    ctx = daemon.DaemonContext(
        stdout=stdout_handle,
    )
    with ctx:
        i = 0
        while True:
            time.sleep(1)
            print(f"record #{i}")
            i += 1
```

I think I'll prepare a PR with truncation included, but I *really* think truncating logs should be a job for logrotate.

**Reply by enoperm** [reply]
See: https://github.com/apache/airflow/pull/25363

**Reply by potiuk** [reply]
Closeed #25363 as we already had https://github.com/apache/airflow/pull/25253 opened

**Reply by enoperm** [reply]
Oh, didn't see that one.
